code/datums/progress_bar.dm

```diff
@@ -0,0 +1,213 @@
+
+	user = param_user
+	goal = param_goal
+	target = param_target
+	init_images(target)
+
+	if(user)
+		client = user.client
+		client?.images |= holder
+		LAZYASSOCLISTADD(user.progress_bars, target, src)
+		animate(holder, pixel_z = world.icon_size + (height * LAZYLEN(user.progress_bars[target]) - 1), alpha = 255, time = animation_time_appear, easing = SINE_EASING | EASE_OUT, flags = ANIMATION_PARALLEL)
+	else
+		animate(holder, alpha = 255, time = animation_time_appear, easing = SINE_EASING | EASE_OUT, flags = ANIMATION_PARALLEL)
+
+/datum/progress_bar/Destroy()
+	client?.images -= holder
+
+	QDEL_NULL(holder)
+	QDEL_NULL(backdrop)
+	QDEL_NULL(fill)
+
+	LAZYASSOCLISTREMOVE(user.progress_bars, target, src)
+
+	target = null
+	user = null
+	client = null
+
+	return ..()
+
+/**
+  * Stops the progress bar and destoy it after a bit.
+  */
+/datum/progress_bar/proc/stop()
+	set waitfor = FALSE
+
+	shift_down_all()
+	update(last_progress)
+	animate(holder, alpha = 0, time = animation_time_fade, flags = ANIMATION_PARALLEL)
+	LAZYASSOCLISTREMOVE(user.progress_bars, target, src)
+
+	sleep(animation_time_fade)
+	qdel(src)
+
+/**
+  * Calls the shift animation on all attached bars.
+  */
+/datum/progress_bar/proc/shift_down_all()
+	if(user?.progress_bars[target])
+		var/list/bars = user.progress_bars[target]
+		for(var/i = bars.Find(src), i <= length(bars), i++)
+			var/datum/progress_bar/P = bars[i]
+			P.shift()
+
+/**
+  * Initializes the visual elements.
+  *
+  * Arguments:
+  * * target - The atom to parent the images to.
+  */
+/datum/progress_bar/proc/init_images(target)
+	holder = image('icons/effects/effects.dmi', target, "nothing", HUD_ABOVE_ITEM_LAYER)
+	holder.mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	holder.alpha = 0
+	holder.plane = HUD_PLANE
+	holder.appearance_flags = KEEP_TOGETHER | APPEARANCE_UI_IGNORE_ALPHA
+
+	backdrop = image(icon, target, "backdrop")
+
+	fill = image(icon, target, "fill")
+	fill.filters = filter(type = "alpha", icon = icon(icon, "mask"), x = -fill_width)
+
+	holder.overlays += backdrop
+	holder.overlays += fill
+
+/**
+  * Call periodically to update the progress bar's progress.
+  *
+  * Arguments:
+  * * progress - The new value, divided with the given goal on creation to calculate the fraction.
+  */
+/datum/progress_bar/proc/update(progress)
+	if(QDELETED(target))
+		qdel(src)
+		return
+
+	if(user?.client != client)
+		client?.images -= holder
+		if(!QDELETED(user) && user?.client)
+			client = user.client
+			client.images += holder
+
+	. = clamp(progress / goal, 0, 1)
+	last_progress = progress
+	if(fill)
+		holder.overlays -= fill
+		update_fill(.)
+		holder.overlays += fill
+
+/**
+  * Updates the fill. Do your visual changes here.
+  *
+  * Arguments:
+  * * fraction - The new (value / goal) fraction.
+  */
+/datum/progress_bar/proc/update_fill(fraction)
+	UNLINT(fill?.filters[1]?.x = -fill_width * (1 - fraction))
+
+/**
+  * Shifts the progress bar down.
+  */
+/datum/progress_bar/proc/shift()
```

Comment:
Why is this a separate proc?

Reply:
For overriding.

Reply:
In what circumstance would you override it?

Reply:
If you want to make a new bar theme.